File: src/shared/_types/equipment-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWeightedTotalStats = void 0;
function getWeightedTotalStats(set) {
    // TODO weighting
    return !set ? 0 : ap(set) + vp(set) + hp(set) + mp(set);
}
exports.getWeightedTotalStats = getWeightedTotalStats;
function ap(set) {
    return set.weapon.ap + set.shield.ap + set.helmet.ap + set.armour.ap + set.accessory.ap;
}
function vp(set) {
    return set.weapon.vp + set.shield.vp + set.helmet.vp + set.armour.vp + set.accessory.vp;
}
function hp(set) {
    return set.weapon.hp + set.shield.hp + set.helmet.hp + set.armour.hp + set.accessory.hp;
}
function mp(set) {
    return set.weapon.mp + set.shield.mp + set.helmet.mp + set.armour.mp + set.accessory.mp;
}

File: backend/src/main/kotlin/dl/equipmentCalculator/model/EquipmentSet.kt
package dl.equipmentCalculator.model

data class EquipmentSet(
        val weapon: Equipment,
        val shield: Equipment,
        val helmet: Equipment,
        val armour: Equipment,
        val accessory: Equipment) {

    companion object {
        fun EquipmentSet?.getWeightedTotalStats(
                apWeight: Int = 0,
                vpWeight: Int = 0,
                hpWeight: Int = 0,
                mpWeight: Int = 0
        ): Int {
            if (this === null) return 0
            return if (apWeight + vpWeight + hpWeight + mpWeight == 0)
                ap + vp + hp + mp else
                ap * apWeight + vp * vpWeight + hp * hpWeight + mp * mpWeight
        }

        private val EquipmentSet.ap: Int
            get() = weapon.ap + shield.ap + helmet.ap + armour.ap + accessory.ap
        private val EquipmentSet.vp: Int
            get() = weapon.vp + shield.vp + helmet.vp + armour.vp + accessory.vp
        private val EquipmentSet.hp: Int
            get() = weapon.hp + shield.hp + helmet.hp + armour.hp + accessory.hp
        private val EquipmentSet.mp: Int
            get() = weapon.mp + shield.mp + helmet.mp + armour.mp + accessory.mp
    }
}
